src/util/sandbox/fd_sandbox_linux.c

```diff
@@ -0,0 +1,281 @@
+}
+
+int
+fd_sandbox_linux( const fd_sandbox_profile_linux_t * profile ) {
+  if ( FD_UNLIKELY( !profile->initialized ) ) {
+    FD_LOG_ERR(( "the sandbox profile has not been initialized" ));
+  }
+
+  // Get target user before doing other kinds of restrictions
+  struct passwd * uinfo = getpwnam( profile->user );
+  if ( FD_UNLIKELY( !uinfo ) ) {
+    FD_LOG_ERR(( "getpwnam: %s", strerror( errno ) ));
+  }
+
+  // todo: maybe not clearenv but only keep vars prefixed with FD_
+  clearenv();
+  fd_sandbox_setup_netns( profile->netns );
+  fd_sandbox_setup_mountns( profile->chroot_path );
+  fd_sandbox_set_resource_limits( profile->max_open_fds );
+  fd_sandbox_change_user( uinfo );
+  fd_sandbox_close_fds_beyond( profile->close_fds_beyond );
+  fd_sandbox_drop_ambient_caps();
+  fd_sandbox_seccomp( profile->seccomp_prog );
+
+  FD_LOG_NOTICE(( "thread group sandboxed" ));
+  return 0;
+}
+
+void
+fd_sandbox_change_user( struct passwd * u ) {
+    
+  if ( FD_UNLIKELY( u->pw_uid == 0 ) ) {
+    FD_LOG_ERR(( "will not run as root" ));
+  }
+
+  FD_LOG_INFO(( "switching to user='%s' uid='%d' gid='%d'", u->pw_name, u->pw_uid, u->pw_gid ));
+
+  // setregid before setreuid otherwise setregid won't be allowed
+
+  if ( FD_UNLIKELY( setregid( u->pw_gid, u->pw_gid ) ) ) {
+    FD_LOG_ERR(( "setregid: %s", strerror( errno ) ));
+  }
+
+  if ( FD_UNLIKELY( setreuid( u->pw_uid, u->pw_uid ) ) ) {
+    FD_LOG_ERR(( "setreuid: %s", strerror( errno ) ));
+  }
+}
+
+void
+fd_sandbox_close_fds_beyond( uint target_max_fd ) {
+  FD_LOG_INFO(( "closing all fds beyond %d", target_max_fd ));
+  long max_fds = sysconf( _SC_OPEN_MAX );
+  for ( long fd = max_fds - 1; fd > target_max_fd; fd-- ) {
+     close( (int)fd );
+  } 
+}
+
+void
+fd_sandbox_set_resource_limits(uint max_open_fds) {
+  FD_LOG_INFO(( "setting resource limits" ));
+  struct rlimit l = {
+    .rlim_cur = max_open_fds,
+    .rlim_max = max_open_fds,
+  };
+
+  if ( FD_UNLIKELY( setrlimit(RLIMIT_NOFILE, &l) == -1 ) ) {
+    FD_LOG_ERR(( "setrlimit: %s", strerror( errno ) ));
+  }
+}
+
+void
+fd_sandbox_setup_netns( char * ns_name ) {
+  char netns_path[ PATH_MAX ];
+  int nsfd;
+
+  FD_LOG_INFO(( "setting up network namespace" ));
+
+  // realize the namespace path
+  int reslen = snprintf( netns_path, ARRAY_SIZE(netns_path), "/var/run/netns/%s", ns_name );
+  if ( FD_UNLIKELY( (ulong) (reslen + 1) > ARRAY_SIZE(netns_path) ) ) {
+    FD_LOG_ERR(( "namespace name too long" ));
+  }
+
+  // get a reference to the namespace
+  if ( FD_UNLIKELY( ( nsfd = open( netns_path, 0 ) ) == -1 ) ) {
+    FD_LOG_ERR(( "netns open: %s", strerror( errno ) ));
+  }
+
+  // enter the namespace
+  if ( FD_UNLIKELY( setns( nsfd, CLONE_NEWNET ) == -1 ) ) {
+    FD_LOG_ERR(( "netns setns: %s", strerror( errno ) ));
+  };
+
+  // close the namespace reference
+  close( nsfd );
+}
+
+void fd_sandbox_drop_ambient_caps( void ) {
+  if ( FD_UNLIKELY( prctl( PR_CAP_AMBIENT, PR_CAP_AMBIENT_CLEAR_ALL, 0, 0, 0 ) ) ) {
```

Comment:
Don't you also need to remove all effective caps from the current process as well?

Reply:
I'm revisiting this whole thing. Thanks for pointing it out! I think that we need to sort out the tile zero issue out first before we can handle caps during init properly.